spec/controllers/two_factor_authentication/otp_expired_controller_spec.rb

```diff
@@ -0,0 +1,46 @@
+require 'rails_helper'
+
+describe TwoFactorAuthentication::OtpExpiredController do
+  describe '#show' do
+    context 'renders' do
+      it 'receives the expected value' do
+        user = build(:user, otp_delivery_preference: 'voice')
+
+        allow(controller).to receive(:otp_expired?).and_return(true)
+
+        stub_sign_in_before_2fa(user)
+
+        get :show
+
+        expect(assigns(:otp_delivery_preference)).to eq('voice')
+      end
+
+      it 'once OTP expires' do
+        user = build(:user, otp_delivery_preference: 'voice', direct_otp_sent_at: Time.zone.now)
+
+        allow(controller).to receive(:otp_expired?).and_return(true)
+        stub_sign_in_before_2fa(user)
+
+        get :show
+
+        expect(response).to render_template(:show)
+      end
+    end
+
+    context 'does not render' do
+      it 'when user is signed out' do
+        allow(controller).to receive(:user_signed_in?).and_return(false)
+
+        get :show
+
+        expect(response).to_not render_template(:show)
+      end
+
+      it 'when otp is still valid' do
+        get :show
```

Comment:
I don't think the test is set up to have a valid OTP like described in the `it` block, so I don't think we're covering the logic of the controller to handle the valid OTP.

I think we could stub it like you did above:

```suggestion
      it 'when otp is still valid' do
        user = build(:user, otp_delivery_preference: 'voice', direct_otp_sent_at: Time.zone.now)
        stub_sign_in_before_2fa(user)

        get :show
```